include/seqan3/search/fm_index/fm_index.hpp

```diff
@@ -388,105 +388,76 @@ class fm_index
         return size() == 0;
     }
 
-    // operator== not implemented by sdsl indices yet
-    // bool operator==(fm_index const & rhs) const noexcept
-    // {
-    //     return index == rhs.index;
-    // }
-
-    // operator== not implemented by sdsl indices yet
-    // bool operator!=(fm_index const & rhs) const noexcept
-    // {
-    //     return !(*this == rhs);
-    // }
-
-    /*!\brief Returns a seqan3::fm_index_cursor on the index that can be used for searching.
-     *        \if DEV
-     *            Cursor is pointing to the root node of the implicit suffix tree.
-     *        \endif
-     * \returns Returns a (unidirectional) seqan3::fm_index_cursor on the index.
+    /*!\brief Compares two indices.
+     * \returns `true` if the indices are equal, false otherwise.
      *
      * ### Complexity
      *
-     * Constant.
+     * Linear.
      *
      * ### Exceptions
      *
      * No-throw guarantee.
      */
-    cursor_type begin() const noexcept
+    bool operator==(fm_index const & rhs) const noexcept
     {
-        return {*this};
+        return (index == rhs.index) && (text_begin == rhs.text_begin) && (text_begin_ss == rhs.text_begin_ss) &&
+               (text_begin_rs == rhs.text_begin_rs);
```

Comment:
What is being compared? Why is it linear?

Reply:
We're comparing the underlying int_vector representation and the test begin positions.
Since comparing the vectors needs to compare the elements, the complexity is linear relative to its size, no?

Reply:
True. But do you need to compare the actual index? Aren't two indexes of the same text identical?

Reply:
Mhm, then I would have to compare the texts. Still linear?
And they could use different traits I guess. Then the indices would be different for the same texts.

Reply:
Well, my thought was, that 
  * if the indexes are guaranteed to be the same for the same text
  * and you store a pointer to the text
  * and the pointer value is identical
→ the index is identical and can be compared in `O(1)`. 

You could opt for a less strict definition of identity and compare the contents of the text instead. Or even say that the text is irrelevant and compare the contents of the index, but from my POV that would be unintuitive, because indexes over different alphabets could be equal if they have the same composition of ranks.

Reply:
Mhm, the problem is that 

1) You could have the same text twice and construct an index each and the indices would not be equal though they are.
2) The text is only a pointer, i.e., when I load an index, the text is a `nullptr` - so I can't really compare the text.

Reply:
> You could have the same text twice and construct an index each and the indices would not be equal though they are.

Yes, if you want this behaviour you need to compare the texts.

> The text is only a pointer, i.e., when I load an index, the text is a nullptr - so I can't really compare the text.

Is the index in a valid state if the pointer is a nullptr?

Reply:
The index is.
The only thing that doesn't work is to ask the `fm_index_cursor` or such for the `query()` since it needs text access for it.
The search works as expected since it only returns text positions and doesn't access the indexed text.

Reply:
Btw

>because indexes over different alphabets could be equal if they have the same composition of ranks.

the `fm_index` is templatised over the text type. So we would have to have another `operator==` anyway to compare, e.g., `fm_index<dna4_vector>` and `fm_index<std::string>`?

Reply:
> The index is.

Hm, we need to discuss this at the strategy meeting. I thought certain indexes also do in-text verification. That would not work, right?

> the fm_index is templatised over the text type. So we would have to have another operator== anyway to compare e.g., fm_index<dna4_vector> and fm_index<std::string>?

I was just listing different notions of equality above, I don't thing this should really be possible.

Reply:
You cannot compare dna4_vector with std::string, why should you be able to compare indexes of either with each other?

Reply:
> Hm, we need to discuss this at the strategy meeting. I thought certain indexes also do in-text verification. That would not work, right?

Well, you still can, but you need to provide the text. The index doesn't own the text and hence it's not serialised with it - that's what we decided on when the fm_index was implemented.

> You cannot compare dna4_vector with std::string, why should you be able to compare indexes of either with each other?

You started the thing with having indices over different alphabets 😀